### PR TITLE
Adding responseURL to jqXHR

### LINKED ISSUE #4339: Adding responseURL field to jqXHR
### Description ###

The XMLHttpRequest object has the [responseURL property](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseURL), which is very useful to detect redirect and especially where the request got redirected.

jQXHR objects however does not expose this property and since there are no possibility to get the underlying XMLHttpRequest object (unless you override jQuery.ajaxSettings.xhr function), it is not possible for user functions to get the url to which the request was redirected at all.

This feature were already discussed [here](https://bugs.jquery.com/ticket/15173) and [here](https://github.com/jquery/jquery/pull/1615), but the feature were never implemented.

### PR DESCRIPTION
**Pull request by csmadhav**
### Summary ###

Fixes: #4339
I've taken some references from #1615 (which needed to be resurrected)
original PR was from @xKerman
Since recently support for some browsers was dropped in jquery respective changed were also added.
Please ignore the previous PR (closed already)

### Checklist ###

* [x] All authors have signed the CLA at https://cla.js.foundation/jquery/jquery
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes
* [ ] If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com

### CODE REVIEW

File: src/ajax/xhr.js [hunk 1]
```diff
@@ -15,7 +16,10 @@ var xhrSuccessStatus = {
 
 		// File protocol always yields status code 0, assume 200
 		0: 200
-	};
+	},
+	xhrSupported = jQuery.ajaxSettings.xhr();
+
+support.responseURL = !!xhrSupported && ( "responseURL" in xhrSupported );
```
Comment (mgol) [inline src/ajax/xhr.js]: From all of our supported (on `master`) browsers, only IE doesn't support `responseUrl`. Since #4387, we identify IE directly & don't define support tests if only IE is affected.

File: src/ajax/xhr.js [hunk 2]
```diff
@@ -73,12 +77,14 @@ jQuery.ajaxTransport( function( options ) {
...
 								xhr.status,
-								xhr.statusText
+								xhr.statusText,
+								""
 							);
 						} else {
 							complete(
 								xhrSuccessStatus[ xhr.status ] || xhr.status,
 								xhr.statusText,
+								support.responseURL ? xhr.responseURL : "",
```
Comment (mgol) [inline src/ajax/xhr.js]: What would happen if we just passed it through without checking for support? If IE 11 doesn't crash but just passes `undefined` through, that'd be fine for me.

File: test/data/mock.php [hunk 3]
```diff
@@ -215,6 +215,14 @@ protected function cspClean( $req ) {
 		file_put_contents( $this->cspFile, '' );
 		unlink( $this->cspFile );
 	}
+	
+	public function responseURL($req) {
+		if ( isset($req->query["url"]) ) {
+			header( "Location: $req->query[url]" );
```
Comment (mgol) [inline test/data/mock.php]: Does that also set the `302` status code, as code in `middleware-mockserver` does?

File: test/unit/ajax.js [hunk 4]
```diff
@@ -1408,6 +1408,103 @@ QUnit.module( "ajax", {
 		};
 	} );
 
+	ajaxTest( "jQuery.ajax() - responseURL", 13, function( assert ) {
+		function expectedUrl( url ) {
+			return jQuery.support.responseURL ? url : "";
+		}
+
+		/**
+		 * this function generates complete
```
Comment (mgol) [inline test/unit/ajax.js]: Please start sentences with an uppercase letter. Also, we usually use single-line comments (`//`) even for comments spanning multiple lines.

File: test/unit/ajax.js [hunk 5]
```diff
@@ -1408,6 +1408,103 @@ QUnit.module( "ajax", {
...
+		 * absolute URL without any relative path components
+		 * example:
+		 * input: http://localhost:9876/base/test/data/testinit.js/../../../test/data/mock.php?action=responseURL&155821644568982930
+		 * output: http://localhost:9876/base/test/data/mock.php?action=responseURL&155821644568982930
+		 */
+		function removeRelativePathComponents( url ) {
```
Comment (mgol) [inline test/unit/ajax.js]: Why is this function necessary? We usually just use our existing `url` function.

File: test/unit/support.js [hunk 6]
```diff
@@ -56,18 +56,33 @@ testIframe(
...
 		userAgent = window.navigator.userAgent,
 		expectedMap = {
-			edge: {},
-			ie_11: {},
-			chrome: {},
-			safari: {},
-			firefox: {},
-			ios: {}
+			edge: {
+				responseURL:true
```
Comment (mgol) [inline test/unit/support.js]: There should be a space after a colon. But this will be removed anyway as we don't need this support property.

File: test/unit/support.js [hunk 7]
```diff
@@ -56,18 +56,33 @@ testIframe(
...
+			},
+			ios: {
+				responseURL:true
+			}
 		};
 
 	if ( /edge\//i.test( userAgent ) ) {
+		version = parseInt( ( /edge\/([0-9.]+)/i ).exec( navigator.userAgent )[ 1 ] );
 		expected = expectedMap.edge;
+		expected.responseURL = version >= 14;
```
Comment (mgol) [inline test/unit/support.js]: We only support last 2 Edge versions, currently 17 & 18. This is not needed.

Please revert all changes done to this file.

**Review by mgol — COMMENTED** [review]
Thanks for the PR! I added some comments.

File: src/ajax/xhr.js (hunk 1 above)
Comment (csmadhav) [reply src/ajax/xhr.js]: Okay! I was doing it for IE 11 & Edge < 14, removing

File: src/ajax/xhr.js (hunk 2 above)
Comment (csmadhav) [reply src/ajax/xhr.js]: removed this condition as well, IE doesn't crash and passes undefined

File: test/data/mock.php (hunk 3 above)
Comment (csmadhav) [reply test/data/mock.php]: Yes

File: test/unit/ajax.js (hunk 5 above)
Comment (csmadhav) [reply test/unit/ajax.js]: This is needed for the case when there is no redirect, in that case, `responseURL` should be equal to the initial URL only. The URL returned by `url` function contains some relative path components like `../../` (checkout the example input to this function in comments) so what happens is that after getting the response, `responseURL` value doesn't contain `../` and its cleaned because of which I am not able to assert it. help me with alternatives here

File: test/unit/support.js (hunk 7 above)
Comment (csmadhav) [reply test/unit/support.js]: Yes, I didn't know that, fixed now

File: test/unit/ajax.js (hunk 5 above)
Comment (csmadhav) [reply test/unit/ajax.js]: hey @mgol can you suggest anything here?

Comment (mgol) [reply test/unit/ajax.js]: @csmadhav I’m at a conference; I’ll have a look next week unless I get to it before.

Comment (mgol) [reply test/unit/ajax.js]: OK, it seems the function can stay but:
1. I'd rename it to `normalizeUrl`; shorter & still fully conveys the meaning.
1. The implementation seems too complex for what it does, how about:

```js
function normalizeUrl( url ) {
	var finalUrl = url;
	while ( finalUrl.indexOf( '/../' ) > -1 ) {
		finalUrl = normalizedUrl.replace(/\/[^/]+\/\.\.\//g, '/');
	}
	return finalUrl;
}
```

File: test/unit/ajax.js [hunk 8]
```diff
@@ -1408,6 +1408,98 @@ QUnit.module( "ajax", {
...
+			url = cleaned + url.substring( match.index + match[ 0 ].length - 1 );
+			if ( ( url.match( /\.\.\//g ) || [] ).length !== 0 ) {
+				url = removeRelativePathComponents( url );
+			}
+			return url;
+		}
+
+		var successURL = removeRelativePathComponents( url( "mock.php?action=responseURL" ) ),
+			errorURL = "http://example.invalid",
+			redirectAndSuccessURL = url( "mock.php?action=responseURL" ) + "&url=" + encodeURIComponent( successURL ),
```
Comment (mgol) [inline test/unit/ajax.js]: The whole string should be passed to `url`, not just a part of it.

File: test/unit/ajax.js [hunk 9]
```diff
@@ -1408,6 +1408,98 @@ QUnit.module( "ajax", {
...
+			var splits = url.substring( 0, match.index ).split( "/" );
+			cleaned = splits.slice( 0, splits.length - count - 1 ).join( "/" );
+			url = cleaned + url.substring( match.index + match[ 0 ].length - 1 );
+			if ( ( url.match( /\.\.\//g ) || [] ).length !== 0 ) {
+				url = removeRelativePathComponents( url );
+			}
+			return url;
+		}
+
+		var successURL = removeRelativePathComponents( url( "mock.php?action=responseURL" ) ),
```
Comment (mgol) [inline test/unit/ajax.js]: You may use this normalization for _comparison_ but not when triggering a request. The `url` function logic is needed for the two ways we run unit tests (PHP & Node.js-based). You have to use it unaltered when you pass the URL to jQuery but you may normalize the URL _in assertions_, when comparing it to the desired one if it's necessary.

**Review by mgol — CHANGES_REQUESTED** [review]
Sorry for the delay. See my latest comments.

File: test/unit/ajax.js (hunk 5 above)
Comment (csmadhav) [reply test/unit/ajax.js]: Yes, that would solve!

File: src/ajax.js [hunk 10]
```diff
@@ -451,6 +451,8 @@ jQuery.extend( {
 			jqXHR = {
 				readyState: 0,
 
+				responseURL: "",
```
[the remark this answers is not in the document]
Comment (csmadhav) [reply src/ajax.js]: it makes sense to change it to `undefined`

File: src/ajax.js [hunk 11]
```diff
@@ -714,7 +716,7 @@ jQuery.extend( {
 		}
 
 		// Callback for when everything is done
-		function done( status, nativeStatusText, responses, headers ) {
+		function done( status, nativeStatusText, responseURL, responses, headers ) {
```
[the remark this answers is not in the document]
Comment (csmadhav) [reply src/ajax.js]: I've added it in this API for now, I agree that 5 parameters are quite a lot we should really think of some other API, help me here.

**Comment by csmadhav** [comment]
sorry for updating this late.

File: src/ajax/xhr.js [hunk 12]
```diff
@@ -73,13 +73,12 @@ jQuery.ajaxTransport( function( options ) {
 						} else {
 							complete(
 								xhrSuccessStatus[ xhr.status ] || xhr.status,
-								xhr.statusText,
-
-								// For XHR2 non-text, let the caller handle it (gh-2498)
+								xhr.statusText, // For XHR2 non-text, let the caller handle it (gh-2498)
```
Comment (mgol) [inline src/ajax/xhr.js]: This comment shouldn't have been moved, please move it back below.

File: test/unit/ajax.js [hunk 13]
```diff
@@ -1523,6 +1523,93 @@ QUnit.module( "ajax", {
...
+				finalUrl = finalUrl.replace( /\/[^/]+\/\.\.\//g, "/" );
+			}
+			return finalUrl;
+		}
+
+		var successURL = url( "mock.php?action=responseURL" ),
+			errorURL = "http://example.invalid",
+			redirectAndSuccessURL = url( "mock.php?action=responseURL&url=" + encodeURIComponent( successURL ) ),
+			redirectAndErrorURL = url( "mock.php?action=responseURL&url=" + encodeURIComponent( errorURL ) ),
+			jsonpURL = baseURL + "mock.php?action=jsonp&callback=?",
```
Comment (mgol) [inline test/unit/ajax.js]: Is there any reason why this uses `baseURL` directly instead of the `url` function?

File: src/ajax.js [hunk 14]
```diff
@@ -711,7 +711,7 @@ jQuery.extend( {
 		}
 
 		// Callback for when everything is done
-		function done( status, nativeStatusText, responses, headers ) {
+		function done( status, nativeStatusText, responses, headers, responseURL ) {
```
Comment (mgol) [inline src/ajax.js]: Unfortunately, this function signature is public API as that's the shape of `completeCallback` as described here: https://api.jquery.com/jQuery.ajaxTransport/

This is a new parameter at the end so it's not a breaking change. However, we need to think if this is a good API as 5 unnamed parameters is quite a lot.

(from my comment from https://github.com/jquery/jquery/pull/4405#discussion_r305066924)

**Review by mgol — CHANGES_REQUESTED** [review]
See newest comments.

**Comment by mgol** [comment]
Adding the "Needs review" label back to discuss https://github.com/jquery/jquery/pull/4405#discussion_r364510504

File: src/ajax.js (hunk 14 above)
Comment (mgol) [reply src/ajax.js]: FYI: I opened #4634 to make adding `responseURL` possible without adding more parameters here.

**Comment by mgol** [comment]
Closing & re-opening the PR to trigger the EasyCLA check...

**Comment by mgol** [comment]
@csmadhav I know it's been a while, but we finally landed `jQuery.ajaxTransport` tests in #5793 and the object-based API in #4634, so this PR is now unblocked. Would you be interested rebasing over the latest `main` & adding `jQuery.ajaxTransport` tests?

**Comment by mgol** [comment]
Closing in favor of #5826.
